File: daily/2025/2467_Most_Profitable_Path_in_a_Tree.cpp
#include <iostream>
#include <vector>
#include <queue>

using namespace std;

class Solution {
    public:
        int mostProfitablePath(vector<vector<int>>& edges, int bob, vector<int>& amount) {
            int n = edges.size() + 1;
            vector<vector<int>> graph(n);
            for (auto& edge : edges){
                graph[edge[0]].push_back(edge[1]);
                graph[edge[1]].push_back(edge[0]);
            }
            vector<int> path_come_from(n);
            vector<int> visited(n, 0);
            queue<int> que;
            que.push(bob);
            visited[bob] = 1;
            while(!que.empty()){
                int curr = que.front();
                que.pop();
                for (int neighbor : graph[curr]){
                    if (!visited[neighbor]){
                        visited[neighbor] = 1;
                        path_come_from[neighbor] = curr;
                        que.push(neighbor);
                    }
                }
            }
            vector<int> path;
            int curr = 0;
            while(curr != bob){
                path.push_back(curr);
                curr = path_come_from[curr];
            }
            // path.push_back(bob);
            reverse(path.begin(), path.end());
            // for (int i : path) cout << i << " ";
            // cout << endl;
            vector<int> dp(n, INT_MIN);
            dp[0] = amount[0];
            que.push(0);
            int ans = INT_MIN;
            amount[0] = 0;
            amount[bob] = 0;
            for (int bob_pos : path){
                amount[bob_pos] /= 2;
                // cout << "bob is at " << bob_pos << ", profit is " << amount[bob_pos] << " at that point" << endl;
                int curr_candidate = que.size();
                while(curr_candidate--){
                    int prev = que.front();
                    que.pop();
                    for (int neighbor : graph[prev]){
                        if (dp[neighbor] != INT_MIN) continue;
                        // cout << "now search " << neighbor << endl;
                        dp[neighbor] = dp[prev] + amount[neighbor];
                        // cout << "pos = " << neighbor << ", profit = " << dp[neighbor] << endl;
                        que.push(neighbor);
                        if (graph[neighbor].size() == 1) ans = max(ans, dp[neighbor]);
                    }
                }
                amount[bob_pos] = 0;
            }
            while(!que.empty()){
                int curr = que.front();
                que.pop();
                for (int neighbor : graph[curr]){
                    if (dp[neighbor] != INT_MIN) continue;
                    dp[neighbor] = dp[curr] + amount[neighbor];
                    amount[neighbor] = 0;
                    if (graph[neighbor].size() == 1) ans = max(ans, dp[neighbor]);
                    que.push(neighbor);
                }
            }
            return ans;
        }
    };

int main(){
    vector<vector<int>> edges = {{0,1},{1,2},{1,3},{3,4}};
    int bob = 3;
    vector<int> amount = {-2,4,2,-4,6};
    // vector<vector<int>> edges = {{0,1}};
    // int bob = 1;
    // vector<int> amount = {-7280,2350};
    // vector<vector<int>> edges = {{0,1},{1,2},{2,3}};
    // int bob = 3;
    // vector<int> amount = {-5644,-6018,1188,-8502};
    Solution S;
    cout << S.mostProfitablePath(edges, bob, amount) << endl;
    return 0;
}

File: daily_python/2467_Most_Profitable_Path_in_a_Tree.py
import os
import sys
from typing import List
from collections import deque

class Solution:
    def mostProfitablePath(self, edges: List[List[int]], bob: int, amount: List[int]) -> int:
        n = len(amount)
        self.bob_path = []
        self.graph = [[] for _ in range(n)]
        self.visited = [False] * n

        for edge in edges:
            self.graph[edge[0]].append(edge[1])
            self.graph[edge[1]].append(edge[0])

        self.find_bob_path(bob)
        
        ans = float('-inf')
        que = deque([0])
        dp = [float('-inf')] * n
        dp[0] = amount[0]
        amount[bob] = 0
        time = 1

        while que:
            candidate = len(que)
            if time < len(self.bob_path):
                amount[self.bob_path[time]] /= 2
            for _ in range(candidate):
                curr = que.popleft()
                for neighbor in self.graph[curr]:
                    if dp[neighbor] == float('-inf'):
                        dp[neighbor] = dp[curr] + amount[neighbor]
                        que.append(neighbor)
                        if len(self.graph[neighbor]) == 1:
                            ans = max(ans, dp[neighbor])
            if time < len(self.bob_path):
                amount[self.bob_path[time]] = 0
            time += 1
        return int(ans)
    def find_bob_path(self, pos):
        self.bob_path.append(pos)
        self.visited[pos] = True
        if pos == 0:
            return True
        for neighbor in self.graph[pos]:
            if not self.visited[neighbor] and self.find_bob_path(neighbor):
                return True
        self.bob_path.pop()
        return False
    
def readDataSet(filename):
    dataset = []
    with open(filename, 'r') as file:
        content = file.read().strip()
        blocks = content.split('\n\n')
        for block in blocks:
            lines = block.split('\n')
            edges = [list(map(int, edge.split(','))) for edge in lines[0].split('=')[1].strip()[2:-3].split('},{')]
            bob = int(lines[1].split('=')[1].strip()[:-1])
            amount = list(map(int, lines[2].split('=')[1].strip()[1:-2].split(',')))
            dataset.append((edges, bob, amount))
    return dataset

if __name__ == '__main__':
    if len(sys.argv) == 1:
        filename = os.path.splitext(__file__)[0] + '.txt'
    else:
        filename = sys.argv[1]
    dataset = readDataSet(filename)
    results = []
    solution = Solution()
    for edges, bob, amount in dataset:
        results.append(solution.mostProfitablePath(edges, bob, amount))
    for index, result in enumerate(results):
        print(f'Example {index + 1} : {result}')